lib/error-handler.js

```diff
@@ -62,13 +63,18 @@ function handleError (reply, error, cb) {
     return
   }
 
-  const result = func(error, reply.request, reply)
-  if (result !== undefined) {
-    if (result !== null && typeof result.then === 'function') {
-      wrapThenable(result, reply)
-    } else {
-      reply.send(result)
+  try {
+    const result = func(error, reply.request, reply)
+    if (result !== undefined) {
+      if (result !== null && typeof result.then === 'function') {
+        wrapThenable(result, reply)
+      } else {
+        reply.send(result)
+      }
     }
+  } catch (err) {
```

Comment:
This is swallowing the original error. You should likely log what is happening and set it as `cause` of `err`.

Reply:
I'm hesitant to set the cause of `err` to `error` because:

- It'll mean we'll have different behaviour depending on whether users use a synchronous vs asynchronous error handler. The asynchronous behaviour to match is: https://github.com/fastify/fastify/blob/e3a07eaa444d0e769802195816d4e1718c2fc9ea/lib/wrapThenable.js#L26-L29
- It'll change the existing behaviour for people with <2 error handlers. These people currently get the error thrown by the child error handler unmodified, but they'd now be getting it with the `cause` property set to the child's child error.
- It'll potentially overwrite the `cause` property of err if one is already set by the error handler that throws - or we'd have to have different behaviour again which might get confusing. (maybe one solution here is to write to a symbol key on the error)
- We may have difficulties with frozen objects or similar - or would need to worry about cloning them correctly.

I'm less opposed to adding the logging (though would encourage that we also add the same to wrapThenable.js if we do want this). In my head we're still on the 'things are okay' path, in that this is a reasonable set up for people with nested error handlers so would suggest either not logging here, or logging at INFO level (level same as how the default error handler does for responses with a < 500 error code).

**Next steps:** based on the above, please advise whether:
1. Do we still want to add logging? (yes/no)
  a. If yes to 1, should this be at INFO level?  (yes/suggest other level)
  b. If yes to 1, do we also want to add this to wrapThenable.js L27? (yes/no)
2. Do we still want to add an error cause? (yes/no)
  a. If yes to 2, do we want to add this as a symbol property? (yes/no)
  b. If yes to 2, do we also want to add this to wrapThenable.js L27? (yes/no)
  c. If yes to 2b, do we also want to add this to wrapThenable.js L41 and L43? (yes/no)

(to clarify, my current position would be no to 1 and 2 for the reasons above, but if yes then yes to all subquestions)

Reply:
I'm a little bit over the fence with the logging (or just having it alone), especially as I'd prefer to have the root/child's error bubbled up to my error handler rather than just log that I might not be able to control the error with.

I'm not sure if the frozen objects are something we should worry too much about; it should be safe to assume all things thrown extend directly from Error, and all the others are considered somehow an antipattern. 

I'm still over the Option 2. Maybe we can see how it looks like and decide from there?

Reply:
So either we log or we set it as the `.cause` property of the error. The important part is that the error is not swallowed and is surfaced somehow to the end user. Do not use a symbol because `.cause` is special and part of the language. Do not add it to `wrapThenable`.